### ISSUE TITLE
opensubtitles.com PermissionError: "'[Errno 13] Permission denied: '/config/cache/363133393566646631626632663' ~ opensubtitlescom.py@190"

### ISSUE THREAD

**Issue by reyemxela** [morpheus65535/bazarr#2563]
**Describe the bug**
Getting error specifically with opensubtitles.com:
```
Throttling opensubtitlescom for 12 hours, until 24/07/03 14:20, because of: PermissionError. Exception info: "'[Errno 13] Permission denied: '/config/cache/363133393566646631626632663' ~ opensubtitlescom.py@190"
```

- I have VIP status on opensubtitles
- This isn't a filesystem permissions issue, despite the wording of the error. I have `/config` mapped correctly with proper ownership.
- System>Providers>Reset clears the status, but searching again immediately throws the error
- Manually deleting the file in `/config/cache` allows the provider to work again for a while

Not sure when it stops working, whether it's after a certain amount of time, or some other event. But I can sometimes download hundreds of subtitles (working through a backlog) before it breaks.

**To Reproduce**
Honestly not sure. Basically:
1. Have opensubtitles.com as a provider
2. Wait for it to break?

From a few things I saw in the discord, it sounds like there are maybe a few others like myself with this issue, and the common thread seems to be opensubtitles.com.

**Expected behavior**
opensubtitles.com working without breaking

**Software (please complete the following information):**
 - Bazarr: 1.4.3 (specifically v1.4.3-ls256 by linuxserver.io, running via Podman v5.0.2)
 - Radarr: 4.0.6.1805
 - Sonarr: 5.7.0.8882
 - OS: Fedora 40, 6.8.8-300.fc40.x86_64

Let me know if any more info is needed, and feel free to ping me on discord, @alexmeyer

**Comment by halali** [comment]
This is not an issue, you just donr have write access to your cache directory... delete cache directory and restart bazarr...

**Comment by morpheus65535** [comment]
Duplicate from #2502 , no?

**Comment by reyemxela** [comment]
Same symptoms, yes, but I'm not running via WSL. There also wasn't a definite resolution to that issue besides the OP never responding.

As for volume mappings, I have `/config` bind-mounted to a folder in my home directory, on my single OS drive. No cross-device/hardlinking shenanigans going on there.

The error (in my logs and in the issue you referenced) always refer to `shutil.move`ing from `/tmp` (within the container filesystem) to `/config/cache`. The thing is, this error only seems to happen occasionally, and only with opensubtitles.com.

**Comment by morpheus65535** [comment]
What other providers are you using? They may not have cache implemented.

Does a file with that name already exists in cache directory?

Could it be related to how the /config volume binding is done with podman? I'm not familiar with it.

**Comment by morpheus65535** [comment]
Can you provide complete exception stack so I can properly work on that? Thanks!

**Comment by reyemxela** [comment]
I turned on debug logging and was able to make it break again, here's the relevant dump (with user/tokens redacted):

```
2024-07-08 16:46:48,823 - subliminal_patch.providers.opensubtitlescom (7faf5ae11b38) :  DEBUG (opensubtitlescom:191) - No cached token, we'll try to login again.
2024-07-08 16:46:48,825 - urllib3.connectionpool           (7faf5ae11b38) :  DEBUG (connectionpool:1055) - Starting new HTTPS connection (1): api.opensubtitles.com:443
2024-07-08 16:46:49,459 - urllib3.connectionpool           (7faf5ae11b38) :  DEBUG (connectionpool:549) - https://api.opensubtitles.com:443 "POST /api/v1/login HTTP/1.1" 200 355
2024-07-08 16:46:49,460 - subliminal_patch.providers.opensubtitlescom (7faf5ae11b38) :  DEBUG (opensubtitlescom:600) - opensubtitlescom returned a standard response. Logging request/response for debugging purpose.
2024-07-08 16:46:49,460 - subliminal_patch.providers.opensubtitlescom (7faf5ae11b38) :  DEBUG (opensubtitlescom:601) - Request URL: https://api.opensubtitles.com/api/v1/login
2024-07-08 16:46:49,460 - subliminal_patch.providers.opensubtitlescom (7faf5ae11b38) :  DEBUG (opensubtitlescom:602) - Request Headers: {'User-Agent': 'Bazarr/1.4.3', 'Api-Key': 'redacted', 'Content-Type': 'application/json', 'Content-Length': '50'}
2024-07-08 16:46:49,460 - subliminal_patch.providers.opensubtitlescom (7faf5ae11b38) :  DEBUG (opensubtitlescom:603) - Request Body: {"username": "redacted", "password": "redacted"}
2024-07-08 16:46:49,461 - subliminal_patch.providers.opensubtitlescom (7faf5ae11b38) :  DEBUG (opensubtitlescom:604) - Response Status Code: 200
2024-07-08 16:46:49,461 - subliminal_patch.providers.opensubtitlescom (7faf5ae11b38) :  DEBUG (opensubtitlescom:605) - Response Headers: {'Date': 'Mon, 08 Jul 2024 20:46:49 GMT', 'Content-Type': 'application/json; charset=utf-8', 'Content-Length': '355', 'Connection': 'keep-alive', 'X-RateLimit-Remaining-Second': '0', 'X-RateLimit-Limit-Second': '1', 'X-RateLimit-Remaining-Hour': '58', 'X-RateLimit-Limit-Hour': '60', 'RateLimit-Reset': '1', 'RateLimit-Remaining': '0', 'RateLimit-Limit': '1', 'X-Frame-Options': 'SAMEORIGIN', 'X-XSS-Protection': '1; mode=block', 'X-Content-Type-Options': 'nosniff', 'X-Download-Options': 'noopen', 'X-Vip-Consumer': 'false', 'X-Vip-User': 'true', 'X-Vip': 'true', 'Cache-Control': 'max-age=0, private, must-revalidate, s-maxage=600', 'Access-Control-Allow-Origin': '*', 'Access-Control-Allow-Methods': 'GET, HEAD, POST, OPTIONS', 'Access-Control-Allow-Headers': 'Origin, Authorization, Accept, Api-Key, Content-Type, X-User-Agent', 'X-Request-Id': 'c2faf556-6976-42c7-8e1a-8f1458674695', 'X-Runtime': '0.275138', 'X-StackifyID': 'V1|d7a14668-79a4-43cb-a3b7-f832eaffcf38|C98087|CD1', 'X-Kong-Upstream-Latency': '277', 'X-Kong-Proxy-Latency': '1', 'X-Cache-Backend': 'apigw1_8000 rb7', 'Age': '0', 'X-Var-Cache': 'MISS', 'X-Via': 'fw1', 'Accept-Ranges': 'bytes', 'CF-Cache-Status': 'DYNAMIC', 'Report-To': '{"endpoints":[{"url":"https:\\/\\/a.nel.cloudflare.com\\/report\\/v4?s=y7jDC7tGde56GWrOTWgkQvKzBiONE5U%2Bx9iIHb1YEbasn%2FOit2Wh59CchQKsEMLWsbHEYeUUlOtcfMRtoxWOwCDlFSNDRjeexC%2FjCCw9eSHMbDM%2Bbpn7VIhx427xh7l504TaqpkALJ0%3D"}],"group":"cf-nel","max_age":604800}', 'NEL': '{"success_fraction":0,"report_to":"cf-nel","max_age":604800}', 'Strict-Transport-Security': 'max-age=15552000; includeSubDomains; preload', 'Server': 'cloudflare', 'CF-RAY': '8a02ed03cfcb6380-ORD', 'alt-svc': 'h3=":443"; ma=86400'}
2024-07-08 16:46:49,461 - subliminal_patch.providers.opensubtitlescom (7faf5ae11b38) :  DEBUG (opensubtitlescom:606) - Response Body: {"user": {"allowed_translations": 10, "allowed_downloads": 1000, "level": "VIP Member", "user_id": redacted, "ext_installed": false, "vip": true}, "token": "redacted", "status": 200, "base_url": "vip-api.opensubtitles.com"}
2024-07-08 16:46:49,463 - subliminal_patch.providers.opensubtitlescom (7faf5ae11b38) :  DEBUG (opensubtitlescom:600) - opensubtitlescom returned a standard response. Logging request/response for debugging purpose.
2024-07-08 16:46:49,463 - subliminal_patch.providers.opensubtitlescom (7faf5ae11b38) :  DEBUG (opensubtitlescom:601) - Request URL: https://api.opensubtitles.com/api/v1/login
2024-07-08 16:46:49,463 - subliminal_patch.providers.opensubtitlescom (7faf5ae11b38) :  DEBUG (opensubtitlescom:602) - Request Headers: {'User-Agent': 'Bazarr/1.4.3', 'Api-Key': 'redacted', 'Content-Type': 'application/json', 'Content-Length': '50'}
2024-07-08 16:46:49,464 - subliminal_patch.providers.opensubtitlescom (7faf5ae11b38) :  DEBUG (opensubtitlescom:603) - Request Body: {"username": "redacted", "password": "redacted"}
2024-07-08 16:46:49,464 - subliminal_patch.providers.opensubtitlescom (7faf5ae11b38) :  DEBUG (opensubtitlescom:604) - Response Status Code: 200
2024-07-08 16:46:49,464 - subliminal_patch.providers.opensubtitlescom (7faf5ae11b38) :  DEBUG (opensubtitlescom:605) - Response Headers: {'Date': 'Mon, 08 Jul 2024 20:46:49 GMT', 'Content-Type': 'application/json; charset=utf-8', 'Content-Length': '355', 'Connection': 'keep-alive', 'X-RateLimit-Remaining-Second': '0', 'X-RateLimit-Limit-Second': '1', 'X-RateLimit-Remaining-Hour': '58', 'X-RateLimit-Limit-Hour': '60', 'RateLimit-Reset': '1', 'RateLimit-Remaining': '0', 'RateLimit-Limit': '1', 'X-Frame-Options': 'SAMEORIGIN', 'X-XSS-Protection': '1; mode=block', 'X-Content-Type-Options': 'nosniff', 'X-Download-Options': 'noopen', 'X-Vip-Consumer': 'false', 'X-Vip-User': 'true', 'X-Vip': 'true', 'Cache-Control': 'max-age=0, private, must-revalidate, s-maxage=600', 'Access-Control-Allow-Origin': '*', 'Access-Control-Allow-Methods': 'GET, HEAD, POST, OPTIONS', 'Access-Control-Allow-Headers': 'Origin, Authorization, Accept, Api-Key, Content-Type, X-User-Agent', 'X-Request-Id': 'c2faf556-6976-42c7-8e1a-8f1458674695', 'X-Runtime': '0.275138', 'X-StackifyID': 'V1|d7a14668-79a4-43cb-a3b7-f832eaffcf38|C98087|CD1', 'X-Kong-Upstream-Latency': '277', 'X-Kong-Proxy-Latency': '1', 'X-Cache-Backend': 'apigw1_8000 rb7', 'Age': '0', 'X-Var-Cache': 'MISS', 'X-Via': 'fw1', 'Accept-Ranges': 'bytes', 'CF-Cache-Status': 'DYNAMIC', 'Report-To': '{"endpoints":[{"url":"https:\\/\\/a.nel.cloudflare.com\\/report\\/v4?s=y7jDC7tGde56GWrOTWgkQvKzBiONE5U%2Bx9iIHb1YEbasn%2FOit2Wh59CchQKsEMLWsbHEYeUUlOtcfMRtoxWOwCDlFSNDRjeexC%2FjCCw9eSHMbDM%2Bbpn7VIhx427xh7l504TaqpkALJ0%3D"}],"group":"cf-nel","max_age":604800}', 'NEL': '{"success_fraction":0,"report_to":"cf-nel","max_age":604800}', 'Strict-Transport-Security': 'max-age=15552000; includeSubDomains; preload', 'Server': 'cloudflare', 'CF-RAY': '8a02ed03cfcb6380-ORD', 'alt-svc': 'h3=":443"; ma=86400'}
2024-07-08 16:46:49,464 - subliminal_patch.providers.opensubtitlescom (7faf5ae11b38) :  DEBUG (opensubtitlescom:606) - Response Body: {"user": {"allowed_translations": 10, "allowed_downloads": 1000, "level": "VIP Member", "user_id": redacted, "ext_installed": false, "vip": true}, "token": "redacted", "status": 200, "base_url": "vip-api.opensubtitles.com"}
2024-07-08 16:46:49,468 - subliminal_patch.core            (7faf5ae11b38) :  ERROR (core:399) - Unexpected error in provider 'opensubtitlescom': Traceback (most recent call last):
  File "/usr/lib/python3.12/shutil.py", line 886, in move
    os.rename(src, real_dst)
OSError: [Errno 18] Cross-device link: '/tmp/tmphybmfpor' -> '/config/cache/38643263303465326664396333373561383963663832366133373930653533393534643532333562'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/app/bazarr/bin/bazarr/../custom_libs/subliminal_patch/core.py", line 368, in list_subtitles_provider
    results = self[provider].list_subtitles(video, to_request)
              ~~~~^^^^^^^^^^
  File "/app/bazarr/bin/bazarr/../custom_libs/subliminal_patch/core.py", line 298, in __getitem__
    provider.initialize()
  File "/app/bazarr/bin/bazarr/../custom_libs/subliminal_patch/providers/opensubtitlescom.py", line 192, in initialize
    self.login()
  File "/app/bazarr/bin/bazarr/../custom_libs/subliminal_patch/providers/opensubtitlescom.py", line 235, in login
    region.set("oscom_server", self.server_hostname)
  File "/app/bazarr/bin/bazarr/../libs/dogpile/cache/region.py", line 1376, in set
    self.backend.set(key, self._value(value))
  File "/app/bazarr/bin/bazarr/../custom_libs/subzero/cache_backends/file.py", line 27, in set
    self._cache[key] = value
    ~~~~~~~~~~~^^^^^
  File "/app/bazarr/bin/bazarr/../libs/fcache/cache.py", line 267, in __setitem__
    self._write_to_file(filename, value)
  File "/app/bazarr/bin/bazarr/../libs/fcache/cache.py", line 252, in _write_to_file
    rename(tmp, filename)
  File "/usr/lib/python3.12/shutil.py", line 906, in move
    copy_function(src, real_dst)
  File "/usr/lib/python3.12/shutil.py", line 475, in copy2
    copyfile(src, dst, follow_symlinks=follow_symlinks)
  File "/usr/lib/python3.12/shutil.py", line 262, in copyfile
    with open(dst, 'wb') as fdst:
         ^^^^^^^^^^^^^^^
PermissionError: [Errno 13] Permission denied: '/config/cache/38643263303465326664396333373561383963663832366133373930653533393534643532333562'
Traceback (most recent call last):
  File "/usr/lib/python3.12/shutil.py", line 886, in move
    os.rename(src, real_dst)
OSError: [Errno 18] Cross-device link: '/tmp/tmphybmfpor' -> '/config/cache/38643263303465326664396333373561383963663832366133373930653533393534643532333562'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/app/bazarr/bin/bazarr/../custom_libs/subliminal_patch/core.py", line 368, in list_subtitles_provider
    results = self[provider].list_subtitles(video, to_request)
              ~~~~^^^^^^^^^^
  File "/app/bazarr/bin/bazarr/../custom_libs/subliminal_patch/core.py", line 298, in __getitem__
    provider.initialize()
  File "/app/bazarr/bin/bazarr/../custom_libs/subliminal_patch/providers/opensubtitlescom.py", line 192, in initialize
    self.login()
  File "/app/bazarr/bin/bazarr/../custom_libs/subliminal_patch/providers/opensubtitlescom.py", line 235, in login
    region.set("oscom_server", self.server_hostname)
  File "/app/bazarr/bin/bazarr/../libs/dogpile/cache/region.py", line 1376, in set
    self.backend.set(key, self._value(value))
  File "/app/bazarr/bin/bazarr/../custom_libs/subzero/cache_backends/file.py", line 27, in set
    self._cache[key] = value
    ~~~~~~~~~~~^^^^^
  File "/app/bazarr/bin/bazarr/../libs/fcache/cache.py", line 267, in __setitem__
    self._write_to_file(filename, value)
  File "/app/bazarr/bin/bazarr/../libs/fcache/cache.py", line 252, in _write_to_file
    rename(tmp, filename)
  File "/usr/lib/python3.12/shutil.py", line 906, in move
    copy_function(src, real_dst)
  File "/usr/lib/python3.12/shutil.py", line 475, in copy2
    copyfile(src, dst, follow_symlinks=follow_symlinks)
  File "/usr/lib/python3.12/shutil.py", line 262, in copyfile
    with open(dst, 'wb') as fdst:
         ^^^^^^^^^^^^^^^
PermissionError: [Errno 13] Permission denied: '/config/cache/38643263303465326664396333373561383963663832366133373930653533393534643532333562'
2024-07-08 16:46:49,474 - root                             (7faf5ae11b38) :  INFO (get_providers:379) - Throttling opensubtitlescom for 12 hours, until 24/07/09 04:46, because of: PermissionError. Exception info: "'[Errno 13] Permission denied: '/config/cache/386432633034653266643963333' ~ opensubtitlescom.py@235"
```

**Comment by reyemxela** [comment]
I actually just figured some things out, it does seem to be a Podman/SELinux/permission issue, but not the kind I've seen before.

I add `:z` to the end of my volume mapping like I do for all my other *arr containers, but it looks like python's `shutil.move` copies the full SELinux label along with the file, including the container-specific "categories" (such as `s0:c181,c926`).

This means on a container delete/recreate, the new container isn't able to modify the files created by the previous one. Minimal example, running in an alpine container with `-v ./test:/test:z`:

```
/ # apk add exa
OK: 53 MiB in 36 packages
/ # touch /tmp/test1 /tmp/test2
/ # exa -lZ /tmp
.rw-r--r-- 0 root system_u:object_r:container_file_t:s0:c410,c673  8 Jul 22:49 test1
.rw-r--r-- 0 root system_u:object_r:container_file_t:s0:c410,c673  8 Jul 22:49 test2
/ # mv /tmp/test1 /test/test1
/ # python -c 'import shutil; shutil.move("/tmp/test2", "/test/test2")'
/ # exa -lZ /test
.rw-r--r-- 0 root system_u:object_r:container_file_t:s0            8 Jul 22:49 test1
.rw-r--r-- 0 root system_u:object_r:container_file_t:s0:c410,c673  8 Jul 22:49 test2
/ #
```

Now if I were to delete this container and run another one, I wouldn't be able to modify `/test/test2` because the new container would get a new set of random categories.

Normally with a `:z` bind mount, files created in the mount don't get any extra category added to the label, but because these files are first created in `/tmp` (where files DO have labeling) and then moved via `shutil.move`, it becomes a problem.

The workaround I've found is to just bind mount the volume with `:Z` instead of `:z`. This makes the bind mount "private" and re-labels the files correctly on container startup. In this case, I don't need any other containers to share the same bind mount, so this should work fine.

This also explains why fully deleting the cache files from the host side would fix the issue for a while, until the container got recreated as part of my nightly backup process.

So in the end, I guess it isn't really a Bazarr bug, but a weird interaction between Podman, SELinux, and shutil.

**tl;dr** for others with this same setup and issue: bind mount `/config` with `:Z`, not `:z`.

**Comment by morpheus65535** [comment]
Thanks for getting back with the solution. For others getting here, let us know if it help you. If it's the case, I'll create a wiki page for this.

**Comment by eliac** [comment]
Thanks, this was extremely helpful!
